File: everything/src/tools/delete-ghost-post.js
/**
 * Tool handler for deleting a post from Ghost blog
 */
import axios from 'axios';
import crypto from 'crypto';

/**
 * Creates a Ghost Admin API JWT token
 * Uses base64 encoding and HMAC-SHA256 for manual JWT creation
 * 
 * @param {string} keyId - The Ghost Admin API key ID
 * @param {string} keySecret - The Ghost Admin API key secret
 * @returns {string} The JWT token
 */
function createToken(keyId, keySecret) {
    // Create header
    const header = {
        alg: "HS256",
        typ: "JWT",
        kid: keyId
    };

    // Create payload
    const iat = Math.floor(Date.now() / 1000);
    const payload = {
        iat: iat,
        exp: iat + 300,  // Token expires in 5 minutes
        aud: "/admin/"
    };

    // Base64 encode function
    const b64encode = (obj) => {
        return Buffer.from(JSON.stringify(obj))
            .toString('base64')
            .replace(/=+$/, '')
            .replace(/\+/g, '-')
            .replace(/\//g, '_');
    };

    // Base64 encode header and payload
    const headerEncoded = b64encode(header);
    const payloadEncoded = b64encode(payload);

    // Create signature
    const message = `${headerEncoded}.${payloadEncoded}`;
    const key = Buffer.from(keySecret, 'hex');
    const signature = crypto.createHmac('sha256', key)
        .update(message)
        .digest('base64')
        .replace(/=+$/, '')
        .replace(/\+/g, '-')
        .replace(/\//g, '_');

    // Combine all parts
    return `${headerEncoded}.${payloadEncoded}.${signature}`;
}

export async function handleDeleteGhostPost(server, args) {
    try {
        // Validate required parameters
        if (!args.post_id) {
            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify({ error: "Post ID is required" }, null, 2),
                }],
                isError: true
            };
        }

        // Get API credentials
        const apiUrl = process.env.GHOST_API_URL || "https://blog.emmanuelu.com";
        const adminId = "67b2d2824fdabf0001eb99ea";
        const adminSecret = "100eda163e9fea6906fbd7ddc841812c70561b19ab2e6d7b31f844f6ccd7b05d";

        try {
            // Create JWT token
            const token = createToken(adminId, adminSecret);

            // Set up axios instance with retry logic
            const axiosInstance = axios.create({
                timeout: 30000,
                headers: {
                    'Authorization': `Ghost ${token}`,
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            });

            // Add retry logic
            axiosInstance.interceptors.response.use(null, async (error) => {
                const { config } = error;
                if (!config || !config.retry) {
                    return Promise.reject(error);
                }
                
                config.retryCount = config.retryCount || 0;
                
                if (config.retryCount >= config.retry) {
                    return Promise.reject(error);
                }
                
                config.retryCount += 1;
                
                const backoff = config.retryDelay || 1000;
                await new Promise(resolve => setTimeout(resolve, backoff));
                
                return axiosInstance(config);
            });

            // Ghost Admin API endpoint
            const url = `${apiUrl.replace(/\/$/, '')}/ghost/api/admin/posts/${args.post_id}`;

            console.log(`Making request to: ${url}`);
            console.log(`Headers: ${JSON.stringify(axiosInstance.defaults.headers)}`);

            // Make the API request
            const response = await axiosInstance.delete(url, {
                retry: 3,
                retryDelay: 1000
            });
            
            // Ghost returns 204 No Content for successful deletion
            if (response.status === 204) {
                return {
                    content: [{
                        type: 'text',
                        text: JSON.stringify({ 
                            success: true, 
                            message: `Post ${args.post_id} deleted successfully` 
                        }, null, 2),
                    }],
                };
            }
            
            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify(response.data, null, 2),
                }],
            };
        } catch (error) {
            let errorMsg = error.message;
            
            if (error.response) {
                errorMsg += `\nResponse: ${JSON.stringify(error.response.data)}`;
            }
            
            return {
                content: [{
                    type: 'text',
                    text: JSON.stringify({ error: `Network or HTTP error - ${errorMsg}` }, null, 2),
                }],
                isError: true
            };
        }
    } catch (error) {
        return server.createErrorResponse(error);
    }
}

/**
 * Tool definition for delete_ghost_post
 */
export const deleteGhostPostToolDefinition = {
    name: 'delete_ghost_post',
    description: 'Deletes a post from Ghost blog',
    inputSchema: {
        type: 'object',
        properties: {
            post_id: {
                type: 'string',
                description: 'The ID of the post to delete',
            }
        },
        required: ['post_id'],
    },
};

File: src/tools/delete_ghost_post.py
import json
import requests
import os
import base64
import hashlib
import hmac
import time
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

def create_token(key_id: str, key_secret: str) -> str:
    """
    Creates a Ghost Admin API JWT token without using PyJWT.
    Uses base64 encoding and HMAC-SHA256 for manual JWT creation.

    Args:
        key_id (str): The Ghost Admin API key ID
        key_secret (str): The Ghost Admin API key secret

    Returns:
        str: The JWT token
    """
    # Create header
    header = {
        "alg": "HS256",
        "typ": "JWT",
        "kid": key_id
    }

    # Create payload
    iat = int(time.time())
    payload = {
        "iat": iat,
        "exp": iat + 300,  # Token expires in 5 minutes
        "aud": "/admin/"
    }

    # Base64 encode header and payload
    def b64encode(obj):
        return base64.urlsafe_b64encode(json.dumps(obj).encode()).rstrip(b'=').decode('utf-8')

    header_encoded = b64encode(header)
    payload_encoded = b64encode(payload)

    # Create signature
    message = f"{header_encoded}.{payload_encoded}"
    key = bytes.fromhex(key_secret)
    signature = hmac.new(key, message.encode(), hashlib.sha256).digest()
    signature_encoded = base64.urlsafe_b64encode(signature).rstrip(b'=').decode('utf-8')

    # Combine all parts
    token = f"{header_encoded}.{payload_encoded}.{signature_encoded}"
    return token

def delete_ghost_post(post_id: str) -> str:
    """
    Deletes a post from Ghost blog.

    This function supports two usage scenarios:
      1) Deleting a post
      2) Retrieving the tool's metadata by passing post_id="__tool_info__"

    Args:
        post_id (str): The ID of the post to delete

    Returns:
        str: JSON-formatted string containing the response
    """
    if post_id == "__tool_info__":
        info = {
            "name": "delete_ghost_post",
            "description": "Deletes a post from Ghost blog",
            "args": {
                "post_id": {
                    "type": "string",
                    "description": "The ID of the post to delete",
                    "required": True
                }
            }
        }
        return json.dumps(info)

    if not post_id:
        return json.dumps({"error": "Post ID is required"})

    api_url = os.environ.get("GHOST_API_URL", "http://192.168.50.80:8083").rstrip("/")
    admin_id = "67b2d2824fdabf0001eb99ea"
    admin_secret = "100eda163e9fea6906fbd7ddc841812c70561b19ab2e6d7b31f844f6ccd7b05d"

    try:
        # Create JWT token manually
        header = {
            "alg": "HS256",
            "typ": "JWT",
            "kid": admin_id
        }
        iat = int(time.time())
        payload = {
            "iat": iat,
            "exp": iat + 300,  # Token expires in 5 minutes
            "aud": "/admin/"
        }

        # Base64 encode header and payload
        header_encoded = base64.urlsafe_b64encode(json.dumps(header).encode()).rstrip(b'=').decode('utf-8')
        payload_encoded = base64.urlsafe_b64encode(json.dumps(payload).encode()).rstrip(b'=').decode('utf-8')
        
        # Create signature
        message = f"{header_encoded}.{payload_encoded}"
        key = bytes.fromhex(admin_secret)
        signature = hmac.new(key, message.encode(), hashlib.sha256).digest()
        signature_encoded = base64.urlsafe_b64encode(signature).rstrip(b'=').decode('utf-8')
        
        # Combine into final token
        token = f"{header_encoded}.{payload_encoded}.{signature_encoded}"

        session = requests.Session()
        adapter = HTTPAdapter(max_retries=Retry(
            total=3,
            backoff_factor=1,
            status_forcelist=[429, 500, 502, 503, 504]
        ))
        session.mount("http://", adapter)
        session.mount("https://", adapter)

        headers = {
            "Authorization": f"Ghost {token}",
            "Accept": "application/json",
            "Content-Type": "application/json"
        }

        # Ghost Admin API endpoint
        url = f"{api_url}/ghost/api/admin/posts/{post_id}"

        print(f"Making request to: {url}")
        print(f"Headers: {headers}")

        response = session.delete(
            url,
            headers=headers,
            timeout=30
        )
        response.raise_for_status()
        
        # Ghost returns 204 No Content for successful deletion
        if response.status_code == 204:
            return json.dumps({"success": True, "message": f"Post {post_id} deleted successfully"})
        return json.dumps(response.json())
    except requests.exceptions.RequestException as e:
        error_msg = str(e)
        if hasattr(e, 'response') and e.response is not None:
            error_msg += f"\nResponse: {e.response.text}"
        return json.dumps({"error": f"Network or HTTP error - {error_msg}"})
    except Exception as e:
        return json.dumps({"error": f"Unexpected error - {str(e)}"})
    finally:
        session.close()

if __name__ == "__main__":
    print("Testing delete_ghost_post tool...")
    
    # Example post deletion
    print("\nDeleting post...")
    try:
        # Get the most recently published post
        api_url = os.environ.get("GHOST_API_URL", "http://192.168.50.80:8083").rstrip("/")
        list_url = f"{api_url}/ghost/api/admin/posts/"
        list_headers = {
            "Authorization": f"Ghost {create_token('67b2d2824fdabf0001eb99ea', '100eda163e9fea6906fbd7ddc841812c70561b19ab2e6d7b31f844f6ccd7b05d')}",
            "Accept": "application/json"
        }
        posts = requests.get(
            list_url,
            params={"limit": 1},
            headers=list_headers
        )
        print(f"List posts response status: {posts.status_code}")
        posts_json = posts.json()
        print(f"List posts response: {posts_json}")


        if posts_json and len(posts_json.get("posts", [])) > 0:
            post_id = posts_json["posts"][0]["id"]
            print(f"Deleting post with ID: {post_id}")
            result = delete_ghost_post(post_id)
            response = json.loads(result)
            if "error" in response:
                print(f"Error deleting post: {response['error']}")
            else:
                print("Post deleted successfully!")
                print(json.dumps(response, indent=2))
        else:
            print("No posts found to delete")
    except Exception as e:
        print(f"Error during test: {str(e)}")